### PR TITLE
NMS-13575: Added generic way to collect SNMP hardware information

### PR DESCRIPTION
**Pull request by christianpape**
* JIRA: https://issues.opennms.org/browse/NMS-13575

### CODE REVIEW

File: opennms-webapp-rest/src/main/java/org/opennms/web/rest/v2/SnmpMetadataRestService.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import javax.ws.rs.core.Response;
+
+@Component
+@Path("snmpmetadata")
+@Transactional
+public class SnmpMetadataRestService {
+
+    /** The node DAO. */
+    @Autowired
+    private NodeDao m_nodeDao;
```
Comment (fooker) [inline opennms-webapp-rest/src/main/java/org/opennms/web/rest/v2/SnmpMetadataRestService.java:53]:
```suggestion
    private NodeDao nodeDao;
```

File: opennms-webapp/src/main/webapp/hardware/list.jsp [hunk 2]
```diff
@@ -43,9 +43,67 @@
...
+<%@ page import="org.opennms.netmgt.model.OnmsMetaData" %>
+<%@ page import="org.opennms.netmgt.model.snmpmetadata.SnmpMetadataEntry" %>
+<%@ page import="org.opennms.netmgt.model.snmpmetadata.SnmpMetadataObject" %>
+<%@ page import="org.opennms.netmgt.model.snmpmetadata.SnmpMetadataTable" %>
+<%@ page import="org.opennms.netmgt.model.snmpmetadata.SnmpMetadataBase" %>
+<%@ page import="org.opennms.netmgt.model.snmpmetadata.SnmpMetadataValue" %>
 <%@taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
 
 <%!
+    public void printSnmpMetadata(final JspWriter out, final SnmpMetadataBase object) throws Exception {
```
Comment (fooker) [inline opennms-webapp/src/main/webapp/hardware/list.jsp:55]: Split to 4 methods with concrete type for param `object`

**Review by fooker — CHANGES_REQUESTED** [review]

File: integrations/opennms-snmp-metadata-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/SnmpMetadataProvisioningAdapter.java [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+        // retrieve the node
+        final OnmsNode node = nodeDao.get(nodeId);
+        if (node == null) {
+            throw new ProvisioningAdapterException("Failed to return node for given nodeId: " + nodeId);
+        }
+
+        // retrieve primary interface
+        final OnmsIpInterface primaryInterface = node.getPrimaryInterface();
+        if (primaryInterface == null) {
+            throw new ProvisioningAdapterException("Can't find primary interface for nodeId: " + nodeId);
```
[the remark this answers is not in the document]
Comment (christianpape) [reply integrations/opennms-snmp-metadata-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/SnmpMetadataProvisioningAdapter.java:113]: done.

File: integrations/opennms-snmp-metadata-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/SnmpMetadataProvisioningAdapter.java [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        }
+
+        final InetAddress ipAddress = primaryInterface.getIpAddress();
+
+        EventBuilder ebldr = null;
+
+        try {
+            // now get the sysObjectId
+            if (node.getSysObjectId() == null) {
+                LOG.warn("Node {} does not support SNMP. Skipping...", nodeId);
```
[the remark this answers is not in the document]
Comment (christianpape) [reply integrations/opennms-snmp-metadata-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/SnmpMetadataProvisioningAdapter.java:123]: done.

File: integrations/opennms-snmp-metadata-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/SnmpMetadataProvisioningAdapter.java [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+            final OnmsMonitoringLocation location = node.getLocation();
+            final String locationName = (location == null) ? null : location.getLocationName();
+            final SnmpAgentConfig agentConfig = snmpConfigDao.getAgentConfig(ipAddress, locationName);
+
+            for (final Config config : configs) {
+                final SnmpObjId rootOId = SnmpObjId.get(config.getTree());
+
+                final CompletableFuture<List<SnmpResult>> resultFuture = locationAwareSnmpClient.walk(agentConfig, rootOId)
+                        .withDescription("walk" + "_" + config.getName() + "_" + node.getLabel())
+                        .withLocation(node.getLocation() != null ? node.getLocation().getLocationName() : null)
```
[the remark this answers is not in the document]
Comment (christianpape) [reply integrations/opennms-snmp-metadata-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/SnmpMetadataProvisioningAdapter.java:154]: done.

File: integrations/opennms-snmp-metadata-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/SnmpMetadataProvisioningAdapter.java [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+                    .collect(Collectors.toList());
+
+            final OnmsMonitoringLocation location = node.getLocation();
+            final String locationName = (location == null) ? null : location.getLocationName();
+            final SnmpAgentConfig agentConfig = snmpConfigDao.getAgentConfig(ipAddress, locationName);
+
+            for (final Config config : configs) {
+                final SnmpObjId rootOId = SnmpObjId.get(config.getTree());
+
+                final CompletableFuture<List<SnmpResult>> resultFuture = locationAwareSnmpClient.walk(agentConfig, rootOId)
```
[the remark this answers is not in the document]
Comment (christianpape) [reply integrations/opennms-snmp-metadata-provisioning-adapter/src/main/java/org/opennms/netmgt/provision/SnmpMetadataProvisioningAdapter.java:152]: don't think that this makes sense here.

File: opennms-model/src/main/java/org/opennms/netmgt/model/snmpmetadata/SnmpMetadataObject.java [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@XmlRootElement(name = "snmp-metadata-object")
+public class SnmpMetadataObject extends SnmpMetadataBase {
+    @XmlTransient
+    final static Pattern PATTERN_1 = Pattern.compile("[^.]+\\[[^\\[]+\\]|[^.]+|[^.]+$");
```
[the remark this answers is not in the document]
Comment (christianpape) [reply opennms-model/src/main/java/org/opennms/netmgt/model/snmpmetadata/SnmpMetadataObject.java:52]: done.

File: opennms-config-model/src/main/java/org/opennms/netmgt/config/snmpmetadata/Config.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    private String name;
+
+    @XmlAttribute(name = "sysObjectId")
+    private String sysObjectId;
+
+    @XmlAttribute(name = "tree")
+    private String tree;
+
+    @XmlAttribute(name = "context")
+    private String context;
```
[the remark this answers is not in the document]
Comment (christianpape) [reply opennms-config-model/src/main/java/org/opennms/netmgt/config/snmpmetadata/Config.java:53]: done.

File: opennms-model/src/main/java/org/opennms/netmgt/model/snmpmetadata/SnmpMetadataObject.java [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        for (final String p : prefixes) {
+            final Matcher matcher = PATTERN_2.matcher(p);
+            if (matcher.find()) {
+                final String newPrefix = (prefix == null ? "" : prefix + ".") + matcher.group(1);
+                final SnmpMetadataTable table = ((SnmpMetadataObject) structure).addTable(matcher.group(1));
+
+                for (final Map.Entry<String, String> dataEntry : data.entrySet()) {
+                    String key = dataEntry.getKey();
+                    if (!Strings.isNullOrEmpty(newPrefix)) {
+                        if (!dataEntry.getKey().startsWith(newPrefix)) {
```
[the remark this answers is not in the document]
Comment (christianpape) [reply opennms-model/src/main/java/org/opennms/netmgt/model/snmpmetadata/SnmpMetadataObject.java:210]: done.

File: opennms-webapp/src/main/webapp/hardware/list.jsp (hunk 2 above)
Comment (christianpape) [reply opennms-webapp/src/main/webapp/hardware/list.jsp:55]: done.

File: opennms-webapp-rest/src/main/java/org/opennms/web/rest/v2/SnmpMetadataRestService.java (hunk 1 above)
Comment (christianpape) [reply opennms-webapp-rest/src/main/java/org/opennms/web/rest/v2/SnmpMetadataRestService.java:53]: done.

File: opennms-model/src/main/java/org/opennms/netmgt/model/snmpmetadata/SnmpMetadataObject.java [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+                .collect(Collectors.toMap(m -> m.getKey(), m -> m.getValue()));
+        return createStructuredMetaData(new SnmpMetadataObject(context), map, null);
+    }
+
+    private static SnmpMetadataBase createStructuredMetaData(final SnmpMetadataBase structure, final Map<String, String> data, final String prefix) {
+        final Set<String> prefixes = new TreeSet<>();
+
+        if (data.containsKey(prefix)) {
+            if (structure instanceof SnmpMetadataObject) {
+                ((SnmpMetadataObject) structure).addValue(prefix, data.get(prefix));
```
[the remark this answers is not in the document]
Comment (christianpape) [reply opennms-model/src/main/java/org/opennms/netmgt/model/snmpmetadata/SnmpMetadataObject.java:171]: Not possible.

**Review by fooker — APPROVED** [review]

**Comment by mmahacek** [comment]
Is this adapter going to ship with installs by default, or be an optional plugin to install?

**Comment by christianpape** [comment]
> Is this adapter going to ship with installs by default, or be an optional plugin to install?

This issuing to be installed by default.

**Comment by mmahacek** [comment]
> > Is this adapter going to ship with installs by default, or be an optional plugin to install?
> 
> 
> 
> This is going to be installed by default.

Okay. I'm doing some work to cleanup/expand the Provisioning Adapters docs and wanted to check. 
Thanks.

**Review by RangerRick — APPROVED** [review]

**Comment by mmahacek** [comment]
As a heads up, provisioning adapters are moving to the Reference section per [NMS-12999](https://issues.opennms.org/browse/NMS-12999).  I'll be moving the doc page, after this PR merges, as part of https://github.com/OpenNMS/opennms/pull/3693

**Pull request merged**